src/gcp/proto.ts

```diff
@@ -0,0 +1,62 @@
+import { FirebaseError } from "../error";
+
+// A proto duration is a number in seconds appended with "s"
+export type Duration = string;
+
+export function secondsFromDuration(d: Duration): number {
+  return +d.slice(0, d.length - 1);
+}
+
+export function durationFromSeconds(s: number): Duration {
+  return `${s}s`;
+}
+
+export function assertOneOf<T>(typename: string, obj: T, oneof: string, ...fields: (keyof T)[]) {
+  const defined = [];
+  for (const key of fields) {
+    const value = obj[key];
+    if (typeof value !== "undefined" && value != null) {
+      defined.push(key);
+    }
+  }
+
+  if (defined.length > 1) {
+    throw new FirebaseError(
+      `Invalid ${typename} definition. ${oneof} can only have one field defined, but found ${defined.join(
+        ","
+      )}`
+    );
+  }
+}
+
+// Utility function to help copy fields from type A to B.
+// As a safety net, catches typos or fields that aren't named the same
```

Comment:
> catches typos or fields that aren't named the same

Does it? it seems like this just silently does nothing in that case

Reply:
Yup. It's turned typos into compiler errors for me already. `keyof` is magical.